src/examples/rpp/doxygen/retry_when.cpp

```diff
@@ -0,0 +1,63 @@
+#include <rpp/rpp.hpp>
+
+#include <iostream>
+#include <string>
+
+/**
+ * @example retry_when.cpp
+ **/
+
+int main()
+{
+    //! [retry_when delay]
+    size_t retry_count = 0;
+    rpp::source::create<std::string>([&retry_count](const auto& sub) {
+        if (retry_count != 4)
+        {
+            sub.on_error({});
```

Comment:
IIUC, the fundamental guarantees of observables are broken in this example:

After `on_error`, no other message must be emitted. This includes the emission of `on_completed`. See [the observable contract](https://reactivex.io/documentation/contract.html) for reference.

> **OnNext**
>    conveys an item that is emitted by the Observable to the observer
> **OnCompleted**
>    indicates that the Observable has completed successfully and that it will be emitting no further items
> **OnError**
>    indicates that the Observable has terminated with a specified error condition and that it will be emitting no further items 

I am having difficulty understanding how the semantics of `retry_when` fit into this contract. Operators usually act on the *output* of an observable. In this example, we have a restart functionality that *backpropagates* into the generator of a cold observable and acts on the observer itself. I am not happy with that design. 

Please explain why we need `retry_when`.

Reply:
Actually not due to as required no any other messages emitted from **this** observable, but they would be emitted from new one (after resubscribe). Any observer obtained on_error wouldn't obtain any new messages. 
Actually we have chain like this
`observable->observer_inside_retry_when->final_observer`
when observable emits error, then only `observer_inside_retry_when` actually obtains error. Instead of forwarding error to `final_observer` it does attempt to subscribe **new** observer_inside_retry_when (let's say `observer_inside_retry_when_2`) observer to **new** observable_2. so old chain is partially destructed
`observable->observer_inside_retry_when->|     final_observer`
and it becomes like
`observable_2->observer_inside_retry_when_2->final_observer`

No any guarantees are broken in this case.

Why do we need it? For example,  to implement error handling loic with custom delaying. Like this

```c++
rpp::source::create<int>([](const auto& observer)
{
    // some hard job to construct state
     while(true) {
          observer.on_next(std::rand());
          if (get_current_cpu_temp() > 95) { // emulating some issues
              observer.on_error(.....);
          }
     }
    // some hard job to destruct state
})
| rpp::operators::retry_when([](const std::exception_ptr&) {
              return rpp::source::timer(std::chrono::seconds{5}, rpp::schedulers::new_thread{}); // kind of "ok, we are obtained error, it is ok, let's wait 5 seconds to make our cpu cooler"
          })
        | rpp::operators::subscribe(...);
```

Reply:
> no any other messages emitted from this observable, but they would be emitted from new one (after resubscribe)

That is precisely my problem with this operator. This only works and makes sense for cold observables. What happens if this operator is applied on a hot observable? It would subscribe to an observable that has been completed. Can we obtain a compile-time error to avoid this?

Reply:
In this case hot observable should be in “error state”. For example, subjects are cachning last error and remit it on new subscriptions. In this case it would be infinite loop, but it is up to user to control it.
Not sure if it is possible to add compile time error. Anyway it can be easily suppressed just via converting observable to dynamic version (and losing all meta info)

Reply:
Given your arguments, I drop my request for a compile-time error message. Also, I accept your explanation:

The current design means that if the hot observable goes into an error state, you obtain zero emissions on consecutive subscriptions—unless you add a stateful relay like `subject`. This behavior follows the contract.  

I recommend that the difference in behavior for hot and cold observables is explicitly mentioned and explained by example in the documentation of `retry_when` to avoid people running into a pitfall.

Reply:
Yeah, sure, thank you!
@CorentinBT , could you please handle this documentation request?